File: Buoi_3/1.6.3.decimal_to_binary.jl
function decimal_to_binary(n)
    stack = Int[] # Sử dụng mảng như ngăn xếp
    println("Chuyển đổi số thập phân: $n sang nhị phân.")

    if n == 0
        push!(stack, 0)
        println("Đã đẩy phần dư '0' vào ngăn xếp.")
    end

    while n > 0
        remainder = n % 2
        push!(stack, remainder)
        println("Chia $n cho 2, phần dư: $remainder --> Đẩy $remainder vào ngăn xếp")
        n = div(n, 2) # Lấy thương nguyên
        println("Thương mới: $n")
    end

    binary = ""
    println("Chuyển đổi ngăn xếp thành số nhị phân:")
    while !isempty(stack)
        digit = pop!(stack)
        binary *= string(digit)
        println("Lấy $digit từ ngăn xếp và nối vào kết quả.")
    end

    return binary
end

number = 13
binary = decimal_to_binary(number)
println("Số thập phân $number chuyển sang nhị phân là $binary")

File: Buoi_3/1.6.3.decimal_to_binary.cpp
#include <iostream>
#include <stack>
#include <string>

std::string decimal_to_binary(int n) {
    std::stack<int> stack;
    std::cout << "Chuyển đổi số thập phân: " << n << " sang nhị phân.\n";
    if (n == 0) {
        stack.push(0);
        std::cout << "Đã đẩy phần dư '0' vào ngăn xếp.\n";
    }

    while (n > 0) {
        int remainder = n % 2;
        stack.push(remainder);
        std::cout << "Chia " << n << " cho 2, phần dư: " << remainder << " --> Đẩy " << remainder << " vào ngăn xếp\n";
        n = n / 2;
        std::cout << "Thương mới: " << n << "\n";
    }

    std::string binary = "";
    std::cout << "Chuyển đổi ngăn xếp thành số nhị phân:\n";
    while (!stack.empty()) {
        binary += std::to_string(stack.top());
        std::cout << "Lấy " << stack.top() << " từ ngăn xếp và nối vào kết quả.\n";
        stack.pop();
    }
    return binary;
}

int main() {
    int number = 13;
    std::string binary = decimal_to_binary(number);
    std::cout << "Số thập phân " << number << " chuyển sang nhị phân là " << binary << "\n";
    return 0;
}
